task4/src/main/scala/com/bootcamp/task/Homework.scala

```diff
@@ -0,0 +1,129 @@
+package com.bootcamp.task
+
+import scala.annotation.tailrec
+
+object Homework {
+
+	// try to implement min different ways (fold, reduce, recursion)
+	def min(list: List[Int]): Option[Int] = {
+		@tailrec
+		def recurrentMin(min: Int, tail: List[Int]): Int = tail match {
+			case Nil => min
+			case head :: tail => recurrentMin(math.min(head, min), tail)
+		}
+		list match {
+			case Nil => None
+			case current :: tail => Some(recurrentMin(current, tail))
+		}
+	}
+
+	// Implement scanLeft (not using scans ofc)
+	def scanLeft[T](zero: T)(list: List[T])(f: (T, T) => T): List[T] = {
+		list match {
+			case Nil => List(zero)
+			case l =>
+				val (_, history) = l.foldLeft((zero, List(zero))) { case ((acc, history), item) =>
+					val res = f(acc, item)
+					(res, history :+ res)
```

Comment:
It is better to add to head of the list and then reverse. Adding to end of an immutable list is quite expensive

Reply:
Also, you have your acc in the last element of history already

Reply:
Good point, I tend to forget it's not a typical Java ArrayList, where adding to end is much cheaper then to the beginning!

Reply:
Would you please elaborate on "Also, you have your acc in the last element of history already"?

Reply:
you can change `l.foldLeft((zero, List(zero)))` to `l.foldLeft(List(zero))` and get your last res from this history. It would work even better with adding to the beginning of the list

Reply:
Got it, true! Fixed